clients/drcachesim/simulator/caching_device.cpp

```diff
@@ -152,19 +157,54 @@ caching_device_t::request(const memref_t &memref_in)
                 parent->stats->child_access(memref, false, cache_block);
                 parent->request(memref);
             }
+            if (snoop_filter != NULL) {
+                // Update snoop filter, other private caches invalidated on write.
+                snoop_filter->snoop(tag, id, (memref.data.type == TRACE_TYPE_WRITE));
+            }
 
-            // FIXME i#1726: coherence policy
-
+            addr_t victim_tag = cache_block->tag;
             // Check if we are inserting a new block, if we are then increment
             // the block loaded count.
-            if (cache_block->tag == TAG_INVALID) {
+            if (victim_tag == TAG_INVALID) {
                 loaded_blocks++;
-            } else if (inclusive && !children.empty()) {
-                for (auto &child : children) {
-                    child->invalidate(cache_block->tag);
+            } else {
+                if (!children.empty() && inclusive) {
+                    for (auto &child : children) {
+                        child->invalidate(victim_tag, INVALIDATION_INCLUSIVE);
+                    }
+                }
+                if (coherent_cache) {
+                    bool child_holds_tag = false;
+                    if (!children.empty()) {
+                        // We must check child caches to
```

Comment:
s/may/should/?

Took me a few minutes to understand that if a child has the victim tag, from the snoop filter's point of view, this cache still owns it (so it shouldn't clear the bit). Maybe expand the comment a bit to clarify?

Reply:
Added some words to clarify that a cache is responsible for tags contained by its children. Let me know if you think this should be clarified further or in other places.